frame/staking/src/lib.rs

```diff
@@ -960,15 +997,39 @@ pub mod pallet {
 	>;
 
 	/// The map from (wannabe) validator stash key to the preferences of that validator.
+	///
+	/// When updating this storage item, you must also update the `CurrentValidatorsCount`.
 	#[pallet::storage]
 	#[pallet::getter(fn validators)]
 	pub type Validators<T: Config> = StorageMap<_, Twox64Concat, T::AccountId, ValidatorPrefs, ValueQuery>;
 
+	/// A tracker to keep count of the number of items in the `Validators` map.
+	#[pallet::storage]
+	pub type CurrentValidatorsCount<T> = StorageValue<_, u32, ValueQuery>;
+
+	/// The maximum validator count before we stop allowing new validators to join.
+	///
+	/// When this value is not set, no limits are enforced.
+	#[pallet::storage]
+	pub type MaxValidatorsCount<T> = StorageValue<_, u32, OptionQuery>;
+
 	/// The map from nominator stash key to the set of stash keys of all validators to nominate.
+	///
+	/// When updating this storage item, you must also update the `CurrentNominatorsCount`.
```

Comment:
Beyond the scope of this PR: has there been any discussion of adding a map with a counter to substrate? Or is there any substrate precedent for creating a wrapper around the map storage api within a pallet that also uses a counter storage item?

Seems like something along those lines could prevent some footgun where we forget to update the counter.

Reply:
I just saw, `CountedStorageMap` (#9089), that looks like it will be nice here